BaragonService/src/main/java/com/hubspot/baragon/service/elb/ApplicationLoadBalancer.java

```diff
@@ -29,23 +62,401 @@ public ApplicationLoadBalancer(Optional<ElbConfiguration> configuration,
+    DescribeTargetHealthRequest targetHealthRequest = new DescribeTargetHealthRequest()
+        .withTargetGroupArn(targetGroup.getTargetGroupArn());
+    List<TargetHealthDescription> targetHealthDescriptions = elbClient
+        .describeTargetHealth(targetHealthRequest)
+        .getTargetHealthDescriptions();
+
+    boolean instanceIsHealthy = false;
+    int healthyCount = 0;
+
+    for (TargetHealthDescription targetHealthDescription : targetHealthDescriptions) {
+      if (targetHealthDescription.getTargetHealth().getState()
+          .equals(TargetHealthStateEnum.Healthy.toString())) {
+        healthyCount += 1;
+        if (targetHealthDescription.getTarget().equals(target)) {
+          instanceIsHealthy = true;
+        }
+      }
+    }
+
+    return instanceIsHealthy && healthyCount == 1;
+  }
+
+  private void guaranteeHasAllSubnets(Collection<String> subnetIds, LoadBalancer loadBalancer) {
+    Collection<String> subnetsOnLoadBalancer = getSubnetsFromLoadBalancer(loadBalancer);
+    Collection<String> subnetsToAdd = new HashSet<>();
+    for (String subnetId : subnetIds) {
+      if (! subnetsOnLoadBalancer.contains(subnetId)) {
+        subnetsToAdd.add(subnetId);
+      }
+    }
+
+    try {
+      SetSubnetsRequest subnetsRequest = new SetSubnetsRequest()
+          .withLoadBalancerArn(loadBalancer.getLoadBalancerArn())
+          .withSubnets(subnetsToAdd);
+      elbClient.setSubnets(subnetsRequest);
+    } catch (AmazonClientException acexn) {
+      LOG.error("Could not attach subnets {} to load balancer {} due to error",
+          subnetsToAdd, loadBalancer.getLoadBalancerName(), acexn);
+      exceptionNotifier.notify(acexn, ImmutableMap.of(
+          "elb", loadBalancer.getLoadBalancerName(),
+          "subnets", subnetsToAdd.toString()));
+    }
+  }
+
+  /**
+   * After this method completes, every agent in baragonAgents should be associated with
+   * a target in the given target group.
+   *
+   * @param targetGroup group to register in
+   * @param baragonAgents agents to be registered
+   */
+  private void guaranteeHasAllTargets(TargetGroup targetGroup,
+                                      Collection<BaragonAgentMetadata> baragonAgents) {
+    Collection<TargetDescription> targetDescriptions = new HashSet<>();
+    for (BaragonAgentMetadata agent : baragonAgents) {
+      try {
+        if (agent.getEc2().getInstanceId().isPresent()) {
+          if (agentShouldRegisterInTargetGroup(agent.getEc2().getInstanceId().get(), targetGroup)) {
+            String instanceId = agent.getEc2().getInstanceId().get();
+            targetDescriptions.add(new TargetDescription().withId(instanceId));
+            LOG.info("Will register agent {} to target in group {}", agent, targetGroup);
+          } else {
+            LOG.debug("Agent {} is already registered", agent);
+          }
+        } else {
+        /* Is this really an appropriate thing to do here? */
+          throw new IllegalArgumentException(
```

Comment:
I think this is good here since we catch it right below. We want to continue trying to sync the remaining agents (which we do), but this is a rather blocking error for that one agent, so logging at error level + the notify is good to do here.